### PR TITLE
Wt 6816 implement modifier and activity code mapping in patrawin connector

### PR DESCRIPTION
**Pull request by alvinllobrera**

### CODE REVIEW

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java [hunk 1]
```diff
@@ -203,24 +207,9 @@ public PostResult postTime(Request request, TimeGroup timeGroup) {
       return PostResult.PERMANENT_FAILURE.withMessage("User does not exist in Patrawin");
     }
 
-    final Set<String> timeGroupModifiers = getTimeGroupModifiers(timeGroup);
```
Comment (alvinllobrera) [inline src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: Ive extracted this because the method already reached the max length

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java [hunk 2]
```diff
@@ -318,4 +307,60 @@ private String tagUpsertPath() {
...
+
+  private Optional<Integer> getTimeGroupActivityCode(final TimeGroup timeGroup) {
+    final Set<String> timeGroupModifiers = getTimeGroupModifiers(timeGroup);
+    if (timeGroupModifiers.size() > 1) {
+      log.error("All time logs within time group should have same modifier, but got: {}", timeGroupModifiers);
+      return Optional.empty();
+    }
+
+    final String timeGroupModifier = timeGroupModifiers.iterator().next();
+    final String activityCodeStr = StringUtils.isEmpty(timeGroupModifier)
```
Comment (Galya-IT) [inline src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: Let's make `StringUtils.isEmpty` -> `StringUtils.isBlank`

**Comment by Galya-IT** [comment]
@vyshane @alvinllobrera Can someone explain to me where those mappings are going to come for the production deployment? I thought for Patrawin the modifiers / activity codes are numbers of invoices (come from table named `Fakturatextnr`)

**Comment by Galya-IT** [comment]
Oops, closed it by mistake :)

**Comment by Galya-IT** [comment]
The only place I see `TAG_MODIFIER_ACTIVITY_CODE_MAPPING` used is to map the default modifier to an activity code, so basically if we have set the default modifier as the default activity code (as I thought initially it will be), it makes the whole mapping thing redundant.

**Comment by alvinllobrera** [comment]
@Galya-IT we have a `modifier_team` table in our db-service. It contains a list of modifier names for each team. If set, the modifiers will be shown in WT Console when posting time row or time group. These modifier names are intended to be user-friendly (e.g. `Chargeable`, `Planning`,  etc)

When connector receives the posted time, it will use the `TAG_MODIFIER_ACTIVITY_CODE_MAPPING` to check which modifier corresponds to an activity code (e.g `Chargeable` modifier corresponds to activity code `21`).

Basically the idea is to make activity codes (or any special tagging a connected system used) user-friendly in WT Console.

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java (hunk 2 above)
Comment (alvinllobrera) [reply src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: Done

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java [hunk 3]
```diff
@@ -318,4 +307,60 @@ private String tagUpsertPath() {
...
+    final Set<String> timeGroupModifiers = getTimeGroupModifiers(timeGroup);
+    if (timeGroupModifiers.size() > 1) {
+      log.error("All time logs within time group should have same modifier, but got: {}", timeGroupModifiers);
+      return Optional.empty();
+    }
+
+    final String timeGroupModifier = timeGroupModifiers.iterator().next();
+    final String activityCodeStr = StringUtils.isBlank(timeGroupModifier)
+        ? modifierActivityCodeMap.get(defaultModifier)
+        : timeGroupModifier;
```
Comment (Galya-IT) [inline src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: If all modifiers have to be user friendly texts mapped to the activity codes behind the scenes, then maybe this should also be mapped - `modifierActivityCodeMap.get(timeGroupModifier)`?

**Comment by Galya-IT** [comment]
> 
> 
> @Galya-IT we have a `modifier_team` table in our db-service. It contains a list of modifier names for each team. If set, the modifiers will be shown in WT Console when posting time row or time group. These modifier names are intended to be user-friendly (e.g. `Chargeable`, `Planning`, etc)
> 
> When connector receives the posted time, it will use the `TAG_MODIFIER_ACTIVITY_CODE_MAPPING` to check which modifier corresponds to an activity code (e.g `Chargeable` modifier corresponds to activity code `21`).
> 
> Basically the idea is to make activity codes (or any special tagging a connected system used) user-friendly in WT Console.

Thanks for the explanation. So this means that the mappings are hard-coded in the DB as well as in the connector and should be kept in sync?

**Comment by vyshane** [comment]
> Thanks for the explanation. So this means that the mappings are hard-coded in the DB as well as in the connector and should be kept in sync?

Yes. We have an issue WT-6393 for a long term solution that deprecates the manual mappings. For now we still need them.

**Comment by Galya-IT** [comment]
Is this connector not the first one that has activity codes as a DB table in their DB? Does mapping mean that they won't be able to use all of their entries in their activity code table?

**Comment by vyshane** [comment]
> Is this connector not the first one that has activity codes as a DB table in their DB?

Not the first.

> Does mapping mean that they won't be able to use all of their entries in their activity code table?

Usually they don't want to use all of them.

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java (hunk 3 above)
Comment (alvinllobrera) [reply src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: Good spot. Updated PR for the fix

File: README.md [hunk 4]
```diff
@@ -51,6 +52,8 @@ docker run -d \
     -e PATRAWIN_JDBC_URL="jdbc:sqlserver://HOST:PORT;databaseName=DATABASE_NAME;ssl=request;useCursors=true" \
     -e PATRAWIN_DB_USER=dbuser \
     -e PATRAWIN_DB_PASSWORD=dbpass \
+    -e TAG_MODIFIER_ACTIVITY_CODE_MAPPING=modifier_name_1:activity_code_id_1,modifier_name_2:activity_code_id_2 \
```
Comment (Galya-IT) [inline README.md]: If any of those mappings are used now, it would result in an error, because activity codes for Patrawin are numbers.

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java [hunk 5]
```diff
@@ -318,4 +307,60 @@ private String tagUpsertPath() {
...
+                throw new IllegalStateException("Invalid Patrawin modifier to activity code mapping. "
+                    + "Expecting 'modifier:activityCode' format, got: " + tagModifierMapping);
+              }
+              return modifierAndWorkCode;
+            })
+            .collect(Collectors.toMap(
+                modifierWorkCodePair -> modifierWorkCodePair[0],
+                modifierWorkCodePair -> modifierWorkCodePair[1])
+            );
+
```
Comment (Galya-IT) [inline src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: Let's add validation in the init of the connector to each and every mapping validating if it's an activity code existing in the Patrawin DB. Otherwise it may result in an error later.

**Comment by Galya-IT** [comment]
> 
> 
> > Is this connector not the first one that has activity codes as a DB table in their DB?
> 
> Not the first.
> 
> > Does mapping mean that they won't be able to use all of their entries in their activity code table?
> 
> Usually they don't want to use all of them.

Do you have any info from Patrawin what would be the production mapping, because I'm very intrigued how they will pull this off from an Invoice table.

File: src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java (hunk 5 above)
Comment (alvinllobrera) [reply src/main/java/io/wisetime/connector/patrawin/PatrawinConnector.java]: @Galya-IT  
I don't think it is needed on the initialization stage as activity code can be added or changed while connector is running. So there's no guarantee that passing the validation here would mean the activity code is valid the whole time.

I think checking if the activity code is valid during posting time is sufficient.

File: README.md (hunk 4 above)
Comment (alvinllobrera) [reply README.md]: Nope, these are just placeholders. We should put correct values when we deploy

Comment (Galya-IT) [reply README.md]: It will be changed for the production, but anyone who want to debug in dev mode will try to use them. Let's make it easier for them now.

Comment (alvinllobrera) [reply README.md]: Hmmm. Since this is README, I think we should be descriptive as possible (e.g. `activity_code_id_1`)so this will give some hint. Putting numeric values here will also fail if it doesn't exists in Patrawin and might be misunderstood as constant/unchangeable.

How about `numeric_activity_code_id_1`, etc?

Comment (Galya-IT) [reply README.md]: `numeric_activity_code` then, since it won't be an `id`

Comment (alvinllobrera) [reply README.md]: Pushed update

**Comment by vyshane** [comment]
> @vyshane Do you have any info from Patrawin what would be the production mapping, because I'm very intrigued how they will pull this off from an Invoice table. I just want to be sure it's something they are familiar with and are going to use, before we merge the PR.

I'm not sure that I understand your question. I'll try to explain anyway. The activity codes are likely to be different for every Patrawin user. This is the flow of how the activity codes are used:

- Firm using Patrawin creates their activity codes in Patrawin
- They start using WiseTime and want to have (usually a subset) of the codes available when they post time via WiseTime
- They give us a list of activity codes
- Which by themselves aren't very descriptive
- So, we map a description to each code. The description is shown in the WiseTime UI.
- When the user posts time, we send the activity description to Patrawin
- The connector sees the description, maps it to its code, and uses that when sending time to Patrawin
- Patrawin knows what to do with the code

The current process isn't very streamlined, and WT-6393 aims to fix that.
